### PR TITLE
CDAP-3724 Make output of metadata REST APIs consistent

### PR DESCRIPTION
**Pull request by poornachandra**
Build - http://builds.cask.co/browse/CDAP-RBT487-1

Related doc changes - https://github.com/caskdata/cdap/pull/4132

### CODE REVIEW

**Comment by poornachandra** [comment]
Lineage output -
```
{
    "start": 1442863938,
    "end": 1442881938, 
    "relations": [
        {
            "access": "read", 
            "components": [
                "reader"
            ], 
            "data": "stream.default.purchaseStream", 
            "program": "flows.default.PurchaseHistory.PurchaseFlow", 
            "runs": [
                "4b5d7891-60a7-11e5-a9b0-42010af01c4d"
            ]
        }, 
        {
            "access": "unknown", 
            "components": [
                "collector"
            ], 
            "data": "dataset.default.purchases", 
            "program": "flows.default.PurchaseHistory.PurchaseFlow", 
            "runs": [
                "4b5d7891-60a7-11e5-a9b0-42010af01c4d"
            ]
        }
    ], 
    "data": {
        "dataset.default.purchases": {
            "entityId": {
                "id": {
                    "instanceId": "purchases", 
                    "namespace": {
                        "id": "default"
                    }
                }, 
                "type": "datasetinstance"
            }
        }, 
        "stream.default.purchaseStream": {
            "entityId": {
                "id": {
                    "namespace": {
                        "id": "default"
                    }, 
                    "streamName": "purchaseStream"
                }, 
                "type": "stream"
            }
        }
    }, 
    "programs": {
        "flows.default.PurchaseHistory.PurchaseFlow": {
            "entityId": {
                "id": {
                    "application": {
                        "applicationId": "PurchaseHistory", 
                        "namespace": {
                            "id": "default"
                        }
                    }, 
                    "id": "PurchaseFlow", 
                    "type": "Flow"
                }, 
                "type": "program"
            }
        }
    }
}
```

**Comment by poornachandra** [comment]
Search output:
```
[
    {
        "entityId": {
            "id": {
                "applicationId": "PurchaseHistory", 
                "namespace": {
                    "id": "default"
                }
            }, 
            "type": "application"
        }
    }, 
    {
        "entityId": {
            "id": {
                "application": {
                    "applicationId": "PurchaseHistory", 
                    "namespace": {
                        "id": "default"
                    }
                }, 
                "id": "PurchaseFlow", 
                "type": "Flow"
            }, 
            "type": "program"
        }
    }
]
```

**Comment by poornachandra** [comment]
Metadata for a run:
```
[
    {
        "entityId": {
            "id": {
                "namespace": {
                    "id": "default"
                }, 
                "streamName": "purchaseStream"
            }, 
            "type": "stream"
        }, 
        "properties": {}, 
        "scope": "USER", 
        "tags": []
    }, 
    {
        "properties": {
            "key1f": "value1f", 
            "key2F": "value2F"
        }, 
        "scope": "USER", 
        "tags": []
    }, 
    {
        "entityId": {
            "id": {
                "instanceId": "purchases", 
                "namespace": {
                    "id": "default"
                }
            }, 
            "type": "datasetinstance"
        }, 
        "properties": {}, 
        "scope": "USER", 
        "tags": []
    }, 
    {
        "properties": {
            "key1": "value1", 
            "key2": "value2"
        }, 
        "scope": "USER", 
        "tags": [
            "qa-app"
        ]
    }
]
```

**Comment by poornachandra** [comment]
Metadata for a program output:
```
[
    {
        "entityId": {
            "id": {
                "application": {
                    "applicationId": "PurchaseHistory", 
                    "namespace": {
                        "id": "default"
                    }
                }, 
                "id": "PurchaseFlow", 
                "type": "Flow"
            }, 
            "type": "program"
        }, 
        "properties": {
            "key1f": "value1f", 
            "key2F": "value2F"
        }, 
        "scope": "USER", 
        "tags": []
    }
]
```

File: cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchResultRecord.java [hunk 1]
```diff
@@ -23,20 +23,14 @@
  * Represent the Metadata search result record.
  */
 public class MetadataSearchResultRecord {
-  private final Id.NamespacedId id;
-  private final MetadataSearchTargetType type;
+  private final Id.NamespacedId entityId;
```
Comment (henry-cask) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchResultRecord.java]: So, NamespacedId now have a type?

**Comment by poornachandra** [comment]
Metadata properties output:
```
{
    "key1": "value1", 
    "key2": "value2"
}
```

**Comment by poornachandra** [comment]
Metadata tags output:
```
[
    "qa-app", 
    "test-app"
]
```

File: cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchResultRecord.java (hunk 1 above)
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchResultRecord.java]: Yes

File: cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchResultRecord.java [hunk 2]
```diff
@@ -44,26 +38,22 @@ public boolean equals(Object o) {
...
-    return Objects.hash(id, type);
+    return Objects.hash(entityId);
   }
 
   @Override
   public String toString() {
-    return "{" +
-      "id='" + id + '\'' +
-      "type='" + type + '\'' +
+    return "MetadataSearchResultRecord{" +
```
Comment (henry-cask) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchResultRecord.java]: Is the standard toString for CDAP? I always try to return valid JSON String to make it easier to caller.

**Comment by henry-cask** [comment]
Did a pass and add some comments.

**Comment by poornachandra** [comment]
Kafka update output -
```
{
   "previous":{
      "entityId":{
         "type":"application",
         "id":{
            "namespace":{
               "id":"default"
            },
            "applicationId":"PurchaseHistory"
         }
      },
      "scope":"USER",
      "properties":{
         "key2":"value2",
         "key1":"value1"
      },
      "tags":[
         "qa-app",
         "test-app"
      ]
   },
   "changes":{
      "additions":{
         "entityId":{
            "type":"application",
            "id":{
               "namespace":{
                  "id":"default"
               },
               "applicationId":"PurchaseHistory"
            }
         },
         "scope":"USER",
         "properties":{

         },
         "tags":[
            "tag1"
         ]
      },
      "deletions":{
         "entityId":{
            "type":"application",
            "id":{
               "namespace":{
                  "id":"default"
               },
               "applicationId":"PurchaseHistory"
            }
         },
         "scope":"USER",
         "properties":{

         },
         "tags":[

         ]
      }
   },
   "updateTime":1442883836781
}```

File: cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchResultRecord.java (hunk 2 above)
Comment (poornachandra) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchResultRecord.java]: We use default Intellij's `toString()` generator today.

**Comment by henry-cask** [comment]
Format looks good and assuming tests pass let's merge this.

LGTM

**Comment by bdmogal** [comment]
LGTM

**Comment by poornachandra** [comment]
Build is green, merging.

**Pull request merged**